bot/cogs/code_snippets.py

```diff
@@ -0,0 +1,216 @@
+import re
+import textwrap
+from urllib.parse import quote_plus
+
+from aiohttp import ClientSession
+from discord import Message
+from discord.ext.commands import Cog
+
+from bot.bot import Bot
+from bot.utils.messages import wait_for_deletion
+
+
+async def fetch_http(session: ClientSession, url: str, response_format: str, **kwargs) -> str:
+    """Uses aiohttp to make http GET requests."""
+    async with session.get(url, **kwargs) as response:
+        if response_format == 'text':
+            return await response.text()
+        elif response_format == 'json':
+            return await response.json()
+
+
+async def fetch_github_snippet(session: ClientSession, repo: str,
+                               path: str, start_line: str, end_line: str) -> str:
+    """Fetches a snippet from a GitHub repo."""
+    headers = {'Accept': 'application/vnd.github.v3.raw'}
+
+    # Search the GitHub API for the specified branch
+    refs = (await fetch_http(session, f'https://api.github.com/repos/{repo}/branches', 'json', headers=headers)
+            + await fetch_http(session, f'https://api.github.com/repos/{repo}/tags', 'json', headers=headers))
+
+    ref = path.split('/')[0]
+    file_path = '/'.join(path.split('/')[1:])
+    for possible_ref in refs:
+        if path.startswith(possible_ref['name'] + '/'):
+            ref = possible_ref['name']
+            file_path = path[len(ref) + 1:]
+            break
+
+    file_contents = await fetch_http(
+        session,
+        f'https://api.github.com/repos/{repo}/contents/{file_path}?ref={ref}',
+        'text',
+        headers=headers,
+    )
+
+    return await snippet_to_md(file_contents, file_path, start_line, end_line)
+
+
+async def fetch_github_gist_snippet(session: ClientSession, gist_id: str, revision: str,
```

Comment:
If you kept everything inside the cog as methods, then you wouldn't need to keep passing around the session. But whatever, this is okay as-is.

Reply:
I also would prefer these inside the cog. I think module-level functions in cogs should be avoided.

Reply:
I kept the regex as global variables because I saw that other cogs did the same. Is everything else that I did regarding this fine?

Reply:
Yes, looks good.